javalanglib/src/main/scala/java/lang/_String.scala

```diff
@@ -734,6 +734,212 @@ for (cp <- 0 to Character.MAX_CODE_POINT) {
+          // different from indent, we don't add a LF at the end unless there's already one
+          out += "\n"
+        }
+        j += 1
+      }
+      if (trailingNL) out += "\n"
+      out
+    }
+
+  def translateEscapes(): String = {
+    if (thisString == "\\") throw new IllegalArgumentException(s"Illegal escape: `\\(end-of-string)`")
+    else {
+
+      def isOctalDigit(c: Char): scala.Boolean = (c: @switch) match {
+        case '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' => true
+        case _                                             => false
+      }
+      def emplace(str: String, idx: Int, n: Int, that: String): String =
+        str.substring(0, idx) + that + str.substring(idx + n)
+
+      var i   = 0
+      var str = thisString
+      while (i < str.length()) {
+        if (str.charAt(i) == '\\' && i + 1 < str.length()) str.charAt(i + 1) match {
+          // <line-terminator>, so CR(\r), LF(\n), or CRLF(\r\n)
+          case '\r' if i + 2 < str.length() && str.charAt(i + 2) == '\n' =>
+            str = emplace(str, i, 3, "")
+            i -= 1 // step back if we didn't add anything
+          case '\r' | '\n'                                               =>
+            str = emplace(str, i, 2, "")
+            i -= 1 // step back if we didn't add anything
+
+          // normal one char escapes
+          case 'b'  => str = emplace(str, i, 2, "\b")
+          case 't'  => str = emplace(str, i, 2, "\t")
+          case 'n'  => str = emplace(str, i, 2, "\n")
+          case 'f'  => str = emplace(str, i, 2, "\f")
+          case 'r'  => str = emplace(str, i, 2, "\r")
+          case 's'  => str = emplace(str, i, 2, " ")
+          case '"'  => str = emplace(str, i, 2, "\"")
+          case '\'' => str = emplace(str, i, 2, "\'")
+          case '\\' => str = emplace(str, i, 2, "\\")
+
+          // we're parsing octal now, as per JLS-3, we got three cases:
+          // 1) [0-3][0-7][0-7]
+          case a @ ('0' | '1' | '2' | '3')
+            if i + 3 < str.length &&
+               isOctalDigit(str.charAt(i + 2)) && isOctalDigit(str.charAt(i + 3)) =>
+            val codePoint = ((a - '0') * 64) + //
+                            ((str.charAt(i + 2) - '0') * 8) + //
+                            (str.charAt(i + 3) - '0')
+            str = emplace(str, i, 4, valueOf(Character.toChars(codePoint)))
+          // 2) [0-7][0-7]
+          case a if isOctalDigit(a) && i + 2 < str.length() && isOctalDigit(str.charAt(i + 2)) =>
+            val codePoint = ((a - '0') * 8) + (str.charAt(i + 2) - '0')
+            str = emplace(str, i, 3, valueOf(Character.toChars(codePoint)))
+          // 3) [0-7]
+          case a if isOctalDigit(a) =>
+            val codePoint = a - '0'
+            str = emplace(str, i, 2, valueOf(Character.toChars(codePoint)))
+          // bad escape otherwise, this catches everything else including the Unicode ones
+          case bad => throw new IllegalArgumentException(s"Illegal escape: `\\$bad`")
```

Comment:
The Scala 2.11 jobs are failing with
```
[info] compiling 234 Scala sources to /localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalib/.2.11/target/scala-2.11/classes ...
[error] Illegal reference to Scala class scala.scalajs.runtime.package$ in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,879,63)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,879,63)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,879,63)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,879,63)
[error] Illegal reference to Scala class scala.StringContext in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,879,63)
[error] Illegal reference to Scala class scala.collection.immutable.Nil$ in ClassName<java.lang.String> at Position(,0,0)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,879,63)
[error] Illegal reference to Scala class scala.scalajs.runtime.package$ in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.StringContext in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.scalajs.runtime.package$ in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
[error] Illegal reference to Scala class scala.collection.Seq in ClassName<java.lang.String> at Position(file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4598/javalanglib/src/main/scala/java/lang/_String.scala,930,57)
```
Unfortunately we cannot use the `s""` interpolator (nor any other interpolator, for that matter) in Scala 2.11, because they are not optimized away and therefore leave references to the Scala standard library. We have to write
```suggestion
          case bad => throw new IllegalArgumentException("Illegal escape: `\\" + bad + "`")
```
The same applies on line 880.